packages/flutter/lib/src/material/page_transitions_theme.dart

```diff
@@ -37,9 +37,15 @@ class _FadeUpwardsPageTransition extends StatelessWidget {
     return SlideTransition(
       position: _positionAnimation,
       // TODO(ianh): tell the transform to be un-transformed for hit testing
-      child: FadeTransition(
-        opacity: _opacityAnimation,
-        child: child,
+      child: RepaintBoundary(
+        // The above RepaintBoundary separates slide animation from fade
+        // animation so we can apply retained rendering optimization to the fade
+        // animation. Otherwise, the offset change from slide will mix with the
+        // opacity change and our optimization won't be applied.
+        child: FadeTransition(
+          opacity: _opacityAnimation,
+          child: child,
+        ),
       ),
     );
   }
```

Comment:
I don't understand how the OpacityLayer gets retained. Every time the RenderOpacity object paints, i.e. every frame of the fade animation, the OpacityLayer is recreated, which means the engine layer is lost. no?